src/query/models/tags.go

```diff
@@ -0,0 +1,278 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package models
+
+import (
+	"bytes"
+	"fmt"
+	"hash/fnv"
+	"sort"
+	"strings"
+)
+
+var (
+	defaultOptions = NewTagOptions()
+)
+
+// NewTags builds a tags with the given size and tag options.
+func NewTags(size int, opts TagOptions) Tags {
+	if opts == nil {
+		opts = defaultOptions
+	}
+
+	return Tags{
+		// Todo: Pool these
+		Tags: make([]Tag, 0, size),
+		Opts: opts,
+	}
+}
+
+// EmptyTags returns empty tags with a default tag options.
+func EmptyTags() Tags {
+	return NewTags(0, nil)
+}
+
+// ID returns a string representation of the tags.
+func (t Tags) ID() string {
+	var (
+		idLen      = t.IDLen()
+		strBuilder = strings.Builder{}
+	)
+
+	strBuilder.Grow(idLen)
+	for _, tag := range t.Tags {
+		strBuilder.Write(tag.Name)
+		strBuilder.WriteByte(eq)
+		strBuilder.Write(tag.Value)
+		strBuilder.WriteByte(sep)
+	}
+
+	return strBuilder.String()
+}
+
+// IDMarshalTo writes out the ID representation
+// of the tags into the provided buffer.
+func (t Tags) IDMarshalTo(b []byte) []byte {
+	for _, tag := range t.Tags {
+		b = append(b, tag.Name...)
+		b = append(b, eq)
+		b = append(b, tag.Value...)
+		b = append(b, sep)
+	}
+
+	return b
+}
+
+// IDLen returns the length of the ID that would be
+// generated from the tags.
+func (t Tags) IDLen() int {
+	idLen := 2 * t.Len() // account for eq and sep
+	for _, tag := range t.Tags {
+		idLen += len(tag.Name)
+		idLen += len(tag.Value)
+	}
+
+	return idLen
+}
+
+// IDWithExcludes returns a string representation of the tags excluding some tag keys.
+func (t Tags) IDWithExcludes(excludeKeys ...[]byte) uint64 {
+	b := make([]byte, 0, t.Len())
+	for _, tag := range t.Tags {
+		// Always exclude the metric name by default
+		if bytes.Equal(tag.Name, t.Opts.GetMetricName()) {
+			continue
+		}
+
+		found := false
+		for _, n := range excludeKeys {
+			if bytes.Equal(n, tag.Name) {
+				found = true
+				break
+			}
+		}
+
+		// Skip the key
+		if found {
+			continue
+		}
+
+		b = append(b, tag.Name...)
+		b = append(b, eq)
+		b = append(b, tag.Value...)
+		b = append(b, sep)
```

Comment:
nit: In the future you don't  need to create a byte slice and then do a fnv hash of it, you can simply create a fnv and then `Write(..)` to it multiple times and take a final `.Sum64()`.

Reply:
Nice! I think we need to properly consider how we generate IDs in general though, since we may have problems if somehow tags have seperator characters in them, as per [this issue](https://github.com/m3db/m3/issues/1001)

We should probably get on that pretty quickly considering this is becoming a prod use case for many users :\

Reply:
That being said, may as well roll it up here as you suggested

Reply:
Actually, may leave it for now; might need to have a discussion to figure out the best way to handle ID generation here since it's going to be super janky either way